stream_read.py

```diff
@@ -0,0 +1,484 @@
+    """Calculates unit cell parameters.
+
+    Parameters
+    ----------
+    astar : list
+
+        List of float values.
+        (A line beginning with `astar`)
+    bstar : list
+
+        List of float values.
+        (A line beginning with `bstar`)
+    cstar : list
+
+        List of float values.
+        (A line beginning with `cstar`)
+
+    Returns
+    -------
+    a, b, c, alfa, beta, gamma : tuple
+
+        Unit cell parameters.
+    """
+
+    def angle_between(x1, x2):
+        """Auxiliary function for calculating the angle between vectors.
+
+        Parameters
+        ----------
+        x1 : numpy.ndarray
+
+            First vector.
+        x2 : numpy.ndarray
+
+            Second vector.
+
+        Returns
+        -------
+        Angle in degrees.
+        """
+        # magnitude of a vector x1
+        mod1 = np.linalg.norm(x1)
+        # magnitude of a vector x2
+        mod2 = np.linalg.norm(x2)
+        cosine = (np.sum(x1*x2)) / (mod1*mod2)
+        if (cosine > 1.0):
+            cosine = 1.0
+        if (cosine < -1.0):
+            cosine = -1.0
+
+        # angle in degrees
+        return np.rad2deg(np.arccos(cosine))
+
+    # Convert reciprocal -> crystallographic.
+    # Start by converting reciprocal -> cartesian
+    array = np.transpose(np.array([astar, bstar, cstar]))
+    # inverse matrix
+    array = np.linalg.inv(array)
+
+    # magnitude of a vector
+    # Multiplying by 10 for Angstroms.
+    a = np.linalg.norm(array[0, :]) * 10
+    b = np.linalg.norm(array[1, :]) * 10
+    c = np.linalg.norm(array[2, :]) * 10
+    # angle between vectors
+    alpha = angle_between(array[1, :], array[2, :])
+    beta = angle_between(array[0, :], array[2, :])
+    gamma = angle_between(array[0, :], array[1, :])
+
+    return a, b, c, alpha, beta, gamma
+
+
+def search_crystals_parameters(file_name):
+    """Searching crystals parameters in indexing stream file.
+
+    The function parses the file.
+
+    Parameters
+    ----------
+    file_name : Python unicode str (on py3)
+
+        Path to stream file.
+
+    Returns
+    -------
+    crystals : list
+
+        List of crystals dictionaries
+        containing  unit cell details.
+
+    Raises
+    ------
+    FileNotFoundError
+        If no such file.
+    IndexError
+        If there is no path to the stream file.
+    """
+    # flags- Check if the line has already been processed.
+    flags = {"name": False, "begin_crystal": False,
+             "astar": False, "bstar": False, "cstar": False,
+             "lattice_type": False, "centering": False,
+             "unique_axis": False}
+
+    # crystals: Output list of crystals dictionaries
+    # containing  unit cell details.
+    crystals = []
+    chunks_counter = 0  # All crystals in the stream file.
+    # line with the event
+    # ( some stream files do not contain these lines)
+    event_name = ""
+
+    try:
+        with open(file_name) as file:
+            for line in file:
```

Comment:
This parsing section could be made cleaner, but that's not too important for this merge request, just a note for later